docs/designs/flatten_odps_key_value_table.md

```diff
@@ -0,0 +1,107 @@
+# Transform the Column with Key-value Pairs String to Wide Table
+This document describes the design to transform the column with key-value pairs string to a wide ODPS table where
+the values of each key is in a each column.
+
+## Motivation
+Generally, the data developers parse logs to features and the number of features may vary. It is convenient to save features using
+key-value pair string as a column in a table. However, it is difficult to analyze features in the table using SQL syntax. So we need to transform
+the column to a wide table. Then we can analyze each features like calculate the mean and variance for the numeric feature. For example, the key-value column is
+
+| features| label|
+|---------|------|
+| age:19,income:1200,education:Master | 1 |
+| age:23,income:4500,education:Doctor | 0 |
+
+And we need to transform it to 
+
+| age | income | education | label |
+|-----|--------|-----------|-------|
+| 19  | 1200   | Master    | 1 |
+| 23  | 4500   | Doctor    | 0 |
+
+
+## Design Components
+
+### ODPS UDF to Transform Key-value Pairs String to Multiple Values
+ODPS provides UDF for users to define transformation using Python. We choose the UDTF in UDF to implement the transformation because it can 
+match any input parameters in SQL. The UDTF demo is
+
+```python
+#coding:utf-8
+from odps.udf import annotate
+from odps.udf import BaseUDTF
+
+class Explode(BaseUDTF):
+   def process(self, arg):
+       ...
+```
+
+For the UDTF to transform key-value pairs, we can implement a UDTF with multiple arguments. The arguments in the UDTF likes 
+```
+udtf_func(kv_column, append_columns, feature_names, inter_kv_separator, intra_kv_separator)
+```
+The "kv_column" is the column name with key-value pairs strings. \
+The "append_columns" is column names which are directly exported without any transformation. \
+The "feature_names" is the a string with feature names separated by ",". \
+The "inter_kv_separator" is the inter key-value pairs separator. \
+The "intra_kv_separator" is the intra key-value pairs separator.
+
+For the table showed in the motivation section, The SQL with the UDTF is 
+```sql
+SELECT udtf_func(features, label, "age, income, education", ",", ":") as (age, income, education, label)  FROM input_table
+```
+
+The implementation of the UDF is
+```python
+#coding:utf-8
+from odps.udf import annotate
+from odps.udf import BaseUDTF
+
+class Explode(BaseUDTF):
+   def process(self, *args):
```

Comment:
`*args` is hard to understand. Please use parameter names explicitly.

Reply:
The number of args may vary because the number of columns to transform is not fixed. The len(args) is dependent on the number of keys in the column. For example, the len(args)=2 when key-value likes "age:10,education_num:2" and len(args)=3 when key-value likes "age:10,education_num:3,region:US".